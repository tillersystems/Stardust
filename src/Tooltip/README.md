# Tooltip

### Usage

```jsx
import { Tooltip } from '@tillersystems/stardust';
```

<!-- STORY -->

### Properties

- `appearance` - Appearance is used to set the color of the tooltip which can be dark or light.
- `arrow` - Determines if an arrow should be added to the tooltip pointing toward the reference element.
- `boundary` - The boundary preventOverflow modifier. Possible values: "scrollParent", "window", "viewport", or an HTMLElement.
- `children` - Anything that can be rendered: numbers, strings, elements or an array (or fragment). It is the element where a tooltip is hooked to.
- `className` - Class needed by styled components.
- `content` - The content of the tooltip. Along with a string or element, you can use a function that takes the reference element as an argument and returns content.
- `maxWidth` - Determines the maximum width of the tooltip.
- `placement` - Positions the tooltip relative to its reference element.
- `trigger` - The events (each separated by a space) which cause a tooltip to show. Possible values: "mouseenter", "focus", "click", "manual". Use "manual" to only trigger the tippy programmatically.

All of the native [Tippy.js](https://atomiks.github.io/tippyjs/) options can be passed as props. Visit [All Options](https://atomiks.github.io/tippyjs/all-options/) to view the complete table.

| propName     |     propType     |    defaultValue    | isRequired |
| ------------ | :--------------: | :----------------: | :--------: |
| `appearance` |     `string`     |       `dark`       |     -      |
| `arrow`      |      `bool`      |      `false`       |     -      |
| `boundary`   |     `string`     |   `scrollParent`   |     -      |
| `children`   |      `node`      |       `null`       |     -      |
| `className`  |     `string`     |        `''`        |     -      |
| `content`    |     `string`     |        `''`        |     -      |
| `maxWidth`   | `string, number` |       `350`        |     -      |
| `placement`  |     `string`     |       `top`        |     -      |
| `trigger`    |     `string`     | `mouseenter focus` |     -      |

### Example

```jsx
import { Tooltip } from '@tillersystems/stardust';

render() {
  return (
    <Tooltip
      appearance="light"
      arrow
      content="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes sur la période séléctionnée."
      maxWidth={400}
      placement="bottom"
      trigger="click"
      >
        <button>
          Show Tooltip
        </button>
      </Tooltip>
  )
}
```
