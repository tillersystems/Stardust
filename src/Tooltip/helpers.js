import { css } from 'styled-components';

/**
 * getAppearance
 *
 * Helper function to render the right css
 * compared to the appearance given to it.
 *
 * @param {object} palette // The color's pallete of the theme.
 * @param {string} appearance // The button appearance.
 */

export const getAppearance = ({ palette }, appearance) => {
  return {
    dark: css`
      color: ${palette.white};
      background: ${palette.darkBlue};
      box-shadow: none;
      &::before,
      &::after {
        border-color: ${palette.darkBlue} transparent ${palette.darkBlue} transparent;
      }
    `,
    light: css`
      color: ${palette.darkBlue};
      background: ${palette.white};
      box-shadow: 0 0 0 1px ${palette.lightGrey}, 0 2px 16px 0 rgba(0, 0, 0, 0.1);
      &::before {
        border-color: ${palette.white} transparent ${palette.white} transparent;
      }

      &::after {
        border-color: ${palette.lightGrey} transparent ${palette.lightGrey} transparent;
      }
    `,
  }[appearance];
};
