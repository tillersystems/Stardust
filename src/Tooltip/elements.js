import styled, { css } from 'styled-components';

import { getAppearance } from './helpers';

export const Wrapper = styled.div`
  position: relative;
  display: inline-block;
`;

export const Container = styled.div`
  position: absolute;
  z-index: 10;
  width: ${({ width }) => width};

  padding: 0.6rem 1rem;

  border-radius: ${({ theme: { dimensions } }) => dimensions.radius};

  font-size: 1.4rem;

  /* Button Appearance: [default, primary, secondary...] */
  ${({ theme, appearance }) => getAppearance(theme, appearance)};

  top: 100%;
  left: 50%;
  transform: translate3d(-50%, 0, 0);

  ${({ top }) =>
    top &&
    css`
      bottom: 100%;
      top: auto;
    `} &::before {
    content: '';

    display: block;

    position: absolute;
    top: 0;
    left: ${({ arrowPositionX }) => arrowPositionX};
    z-index: -1;

    height: 0.7rem;
    width: 0.7rem;

    transform: translate3d(0, -50%, 0) rotate(135deg);
    ${({ top }) =>
      top &&
      css`
        bottom: 0;
        top: auto;
        transform: translate3d(0, 50%, 0) rotate(135deg);
      `};
  }
`;
