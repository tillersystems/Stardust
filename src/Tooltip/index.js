import React from 'react';
import PropTypes from 'prop-types';
import posed, { PoseGroup } from 'react-pose';

import { Container, Wrapper } from './elements';

/**
 * Tooltip
 *
 * This component is in charge of displaying
 * a tooltip
 *
 * @param {bool} active // Boolean set if the Tooltip is showed or not.
 * @param {bool} hover // Boolean set to use click or hover to show the Tooltip.
 * @param {bool} top // Boolean set to positionate the tooltip above or below the element.
 * @param {string} arrowPositionX // Set position of the Tooltip's arrow.
 * @param {node} children // Anything that can be rendered: numbers, strings, elements or an array (or fragment). It is the element where a tooltip is hooked to.
 * @param {string} width // Tooltip width.
 * @param {string} title // Tooltip main text.
 *
 * @return {jsx}
 */

class Tooltip extends React.Component {
  /** Prop types. */
  static propTypes = {
    active: PropTypes.bool,
    hover: PropTypes.bool,
    top: PropTypes.bool,
    arrowPositionX: PropTypes.string,
    children: PropTypes.node,
    width: PropTypes.string,
    title: PropTypes.string.isRequired,
  };

  /** Default props. */
  static defaultProps = {
    active: false,
    hover: false,
    top: false,
    arrowPositionX: '50%',
    children: null,
    width: 'auto',
  };

  state = {
    active: false,
  };

  /**
   * Handles mounting in component's lifecycle.
   */
  componentDidMount() {
    const { active } = this.props;
    this.setState({ active });
  }

  /**
   * Handles update in component's lifecycle.
   *
   * @param {Object} prevProps - The component's previous props.
   */
  componentDidUpdate(prevProps) {
    const { active } = this.props;
    if (active !== prevProps.active) {
      this.setState({ active });
    }
  }

  handleClick = () => {
    const { active } = this.state;
    this.setState({ active: !active });
  };

  handleMouveEnter = () => {
    this.setState({ active: true });
  };

  handleMouseLeave = () => {
    this.setState({ active: false });
  };

  render() {
    const { top, hover, arrowPositionX, children, width, title } = this.props;
    const { active } = this.state;
    const childrenWithProps = React.Children.map(children, child =>
      React.cloneElement(child, {
        onClick: hover ? null : this.handleClick,
        onMouseEnter: hover ? this.handleMouveEnter : null,
        onMouseLeave: hover ? this.handleMouseLeave : null,
      }),
    );
    return (
      <Wrapper>
        {childrenWithProps}
        <PoseGroup animateOnMount>
          {active && (
            <ToolTipAnimation
              top={top}
              width={width}
              pose={active ? 'enter' : 'exit'}
              arrowPositionX={arrowPositionX}
              key="ContainerAnimation"
            >
              {title}
            </ToolTipAnimation>
          )}
        </PoseGroup>
      </Wrapper>
    );
  }
}

/**
 * Animation
 */
const ToolTipAnimation = posed(Container)({
  enter: {
    x: '-50%',
    y: ({ top }) => (top ? '-15px' : '15px'),
    opacity: 1,
  },
  exit: {
    x: '-50%',
    y: ({ top }) => (top ? '-30px' : '30px'),
    opacity: 0,
  },
});

export default Tooltip;
