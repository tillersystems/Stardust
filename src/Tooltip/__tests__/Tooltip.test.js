import React from 'react';
import { fireEvent } from 'react-testing-library';
import { Theme } from '../..';

import Tooltip from '..';

describe('<Tooltip />', () => {
  test('should render without a problem', () => {
    const { container } = render(
      <Tooltip
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">Show Tooltip</button>
      </Tooltip>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render without a problem with light appearance', () => {
    const { getByTestId } = render(
      <Tooltip
        active
        appearance="light"
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">Show Tooltip</button>
      </Tooltip>,
    );
    const tooltipNode = getByTestId('tooltip');

    expect(tooltipNode).toHaveStyleRule('color', Theme.palette.darkBlue);
    expect(tooltipNode).toHaveStyleRule('background', Theme.palette.white);
    expect(tooltipNode).toHaveStyleRule(
      'box-shadow',
      `0 0 0 1px ${Theme.palette.lightGrey},0 2px 16px 0 rgba(0,0,0,0.1)`,
    );
    expect(tooltipNode).toHaveStyleRule('background', Theme.palette.white, {
      modifier: '::before',
    });
  });

  test('should render an active tooltip without a problem', () => {
    const { getByTestId } = render(
      <Tooltip
        active
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">Show Tooltip</button>
      </Tooltip>,
    );
    const tooltipNode = getByTestId('tooltip');

    expect(tooltipNode).toBeInTheDocument();
  });

  test('should render with a different width', () => {
    const width = '40rem';
    const { getByTestId } = render(
      <Tooltip
        active
        width={width}
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">Show Tooltip</button>
      </Tooltip>,
    );
    const tooltipNode = getByTestId('tooltip');

    expect(tooltipNode).toHaveStyleRule('width', width);
  });

  test('should toggle on click', done => {
    const buttonText = 'Show Tooltip';
    const { queryByTestId, getByTestId, getByText } = render(
      <Tooltip
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">{buttonText}</button>
      </Tooltip>,
    );
    const buttonNode = getByText(buttonText);

    fireEvent.click(buttonNode);

    let tooltipNode;
    tooltipNode = getByTestId('tooltip');

    expect(tooltipNode).toBeInTheDocument();

    fireEvent.click(buttonNode);

    setTimeout(() => {
      tooltipNode = queryByTestId('tooltip');
      expect(tooltipNode).toBeNull();
      done();
    }, 500);
  });

  test('should toggle on mouseEnter/mouseLeave ', done => {
    const buttonText = 'Show Tooltip';
    const { queryByTestId, getByTestId, getByText } = render(
      <Tooltip
        hover
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">{buttonText}</button>
      </Tooltip>,
    );
    const buttonNode = getByText(buttonText);

    fireEvent.mouseEnter(buttonNode);

    let tooltipNode;
    tooltipNode = getByTestId('tooltip');

    expect(tooltipNode).toBeInTheDocument();

    fireEvent.mouseLeave(buttonNode);

    setTimeout(() => {
      tooltipNode = queryByTestId('tooltip');
      expect(tooltipNode).toBeNull();
      done();
    }, 500);
  });

  test('should render a tooltip  displayed above the element', () => {
    const { getByTestId } = render(
      <Tooltip
        active
        top
        title="Ventes nettes (ventes brutes moins les réductions et les annulations) plus les taxes
        sur la période séléctionnée."
      >
        <button type="button">Show Tooltip</button>
      </Tooltip>,
    );
    const tooltipNode = getByTestId('tooltip');

    expect(tooltipNode).toHaveStyleRule('bottom', '100%');
    expect(tooltipNode).toHaveStyleRule('top', 'auto');
  });
});
