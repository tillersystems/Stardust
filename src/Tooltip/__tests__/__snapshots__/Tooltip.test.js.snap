// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should render withouth a problem 1`] = `
.c0 {
  position: relative;
  display: inline-block;
}

<div
  class="c0"
>
  <button
    type="button"
  >
    Show Tooltip
  </button>
</div>
`;
